flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java

```diff
@@ -331,4 +332,103 @@ public class CheckpointingOptions {
+                                                    + "checkpoints to share the same physical file.")
+                                    .build());
+
+    /** The max size of a physical file for merged checkpoints. */
+    @Experimental
+    @Documentation.Section(Documentation.Sections.FILE_MERGING)
+    public static final ConfigOption<MemorySize> FILE_MERGING_MAX_FILE_SIZE =
+            ConfigOptions.key("state.checkpoints.file-merging.max-file-size")
+                    .memoryType()
+                    .defaultValue(MemorySize.parse("32MB"))
+                    .withDescription("Max size of a physical file for merged checkpoints.");
+
+    /**
+     * Whether to use Blocking or Non-Blocking pool for merging physical files. A Non-Blocking pool
+     * will always provide usable physical file without blocking. It may create many physical files
+     * if poll file frequently. When poll a small file from a Blocking pool, it may be blocked until
+     * the file is returned.
+     */
+    @Experimental
+    @Documentation.Section(Documentation.Sections.FILE_MERGING)
+    public static final ConfigOption<Boolean> FILE_MERGING_POOL_BLOCKING =
+            ConfigOptions.key("state.checkpoints.file-merging.pool-blocking")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription(
+                            "Whether to use Blocking or Non-Blocking pool for merging physical files. "
+                                    + "A Non-Blocking pool will always provide usable physical file without blocking. It may create many physical files if poll file frequently. "
+                                    + "When poll a small file from a Blocking pool, it may be blocked until the file is returned.");
+
+    /**
+     * The upper limit of the file pool size based on the number of subtasks within each TM (only
+     * for merging private state at Task Manager level).
+     */
+    @Experimental
+    @Documentation.Section(Documentation.Sections.FILE_MERGING)
+    public static final ConfigOption<Integer> FILE_MERGING_MAX_SUBTASKS_PER_FILE =
+            ConfigOptions.key("state.checkpoints.file-merging.max-subtasks-per-file")
+                    .intType()
+                    .defaultValue(0)
+                    .withDescription(
+                            "The upper limit of the file pool size based on the number of subtasks within each TM"
+                                    + "(only for merging private state at Task Manager level).");
+
+    /**
+     * A threshold that triggers a compaction (re-uploading) of one physical file. If the amount of
+     * invalid data in a physical file exceeds the threshold, a new physical file will be created
+     * and uploaded.
+     */
+    @Experimental
+    @Documentation.Section(Documentation.Sections.FILE_MERGING)
+    public static final ConfigOption<Float> FILE_MERGING_MAX_SPACE_AMPLIFICATION =
```

Comment:
I'd suggest annotate `FILE_MERGING_MAX_SPACE_AMPLIFICATION`, `FILE_MERGING_MAX_SUBTASKS_PER_FILE` and `FILE_MERGING_POOL_BLOCKING` with `@Documentation.ExcludeFromDocumentation`, since these are to be implemented or user are not able to specify currently.